### ISSUE TITLE
@preDestroy hooks are not called on leaf node first

### ISSUE THREAD

**Issue by Luluno01** [remojansen/InversifyJS#1602]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

`@preDestroy` hooks are not called in leaf-parent order.

The order of calling `preDestroy` hooks should be the reverse of the order of they are constructed. That is, we should destroy the parents on the dependency tree first before we destroy their leaves because parents depend on leaves. Destroying leaves first breaks the dependencies of the parents.

NOTE: by dependency tree/graph I mean a directed graph where an edge `parent => child` represents `parent`'s runtime dependence on `child`. This is similar to an object reference graph.

### Steps to reproduce

1. Define two classes `ServiceA` and `ServiceB`, where `ServiceB` depends on `ServiceA` via `@inject(ServiceA)`.
2. `await container.getAsync(ServiceB)`: objects constructed in the correct order.
3. `await container.unbindAllAsync()`: objects destroyed in the wrong order.

```TypeScript
// File: ServiceA.ts
import { injectable, postConstruct, preDestroy } from 'inversify'
import { setTimeout } from 'timers/promises'

@injectable()
export default class ServiceA {
  public constructor() {
    console.log('ServiceA constructor')
  }

  @postConstruct()
  public async init() {
    await setTimeout(1000)
    console.log('ServiceA init')
  }

  @preDestroy()
  public async destroy() {
    await setTimeout(1000)
    console.log('ServiceA destroy')
  }
}
```

```TypeScript
// File: ServiceB.ts
import { inject, injectable, postConstruct, preDestroy } from 'inversify'
import { setTimeout } from 'timers/promises'
import ServiceA from './ServiceA'

@injectable()
export default class ServiceB {
  constructor(
    @inject(ServiceA) public readonly serviceA: ServiceA
  ) {
    console.log('ServiceB constructor')
  }

  @postConstruct()
  public async init() {
    await setTimeout(1000)
    console.log('ServiceB init')
  }

  @preDestroy()
  public async destroy() {
    await setTimeout(1000)
    console.log('ServiceB destroy')
  }
}
```

```TypeScript
// File: inversify.config.ts
import { Container } from 'inversify'
import ServiceA from './ServiceA'
import ServiceB from './ServiceB'

export const container = new Container()
container.bind(ServiceA).toSelf().inSingletonScope()
container.bind(ServiceB).toSelf().inSingletonScope()
```

```TypeScript
// File: index.ts
import 'reflect-metadata/lite'
import { container } from './inversify.config'
import ServiceB from './ServiceB'

await container.getAsync(ServiceB)
/*
ServiceA constructor
ServiceA init
ServiceB constructor
ServiceB init
*/
await container.unbindAllAsync()
/*
ServiceA destroy
ServiceB destroy
*/

export {}
```

In the reproducer above, the dependency graph is `ServiceB => ServiceA`, i.e., `ServiceB` depends on `ServiceA`.

### Expected behavior

`await container.unbindAllAsync()` should always call `preDestroy` hook on the instance of `ServiceB` first, wait until the hook resolves, then call `preDestroy` hook the instance of `ServiceA`.

### Possible solution

Use a tree/graph structure to track the dependencies among singletons. When `unbindAll` or `unbindAllAsync` is called, iteratively call `preDestroy` hooks on the singletons on the dependency graph who have no parent nodes. That is, nodes can be destroyed only when they are root nodes. Destroying root nodes does not break their children.

### Package version

6.0.3

### Node.js version

22.11.0

### In which operating systems have you tested?

- [ ] macOS
- [ ] Windows
- [X] Linux

**Comment by Luluno01** [comment]
Current implementation does not enforce any specific traversal order but rely on the [`Map.prototype.forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/forEach) method, which traverses in insertion order.

https://github.com/inversify/InversifyJS/blob/3761eb2182a11d2693692ebd344ba970d114a5a5/src/container/container.ts#L255-L259

**Comment by Luluno01** [comment]
I'm too lazy to do the mathematical proof of whether the reversed order of [`Map.prototype.forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/forEach) is the exact order we want, but intuitively it should be LOL.

**Comment by notaphplover** [comment]
Hey @Luluno01, I don't see how this is a bug. [Current docs](https://github.com/inversify/InversifyJS/blob/master/wiki/deactivation_handler.md) doesn't say the expected behavior is the one you suggests.

Having said that, we could talk about which behavior could make more sense. This issue is sort of confusing to me. In the first lines you state:

> The order of calling preDestroy hooks should be the reverse of the order of they are constructed. That is, we should destroy the leaves on the dependency tree first before we destroy their parents because leaves depend on parents. Destroying parents first breaks the dependencies of the leaves.

But later, in your example:

> `await container.unbindAllAsync()` should always call `preDestroy` hook on the instance of `ServiceB` first, wait until the hook resolves, then call `preDestroy` hook the instance of `ServiceA`.

In the context of a tree, I understand `ServiceB` is the parent node since `ServiceB` is allowed to have multiple children nodes. I understand parent nodes should be destroyed before children nodes, which makes total sense to me.

Did I understand you properly?

Either way, the current behavior does not enter in conflict with docs and therefore this is not an issue unless I'm missing something. I'll close the issue for now. If, after reading the docs you consider this is an issue, please mention me in this thread and I'll reopen it. Otherwise I would encourage you to open a [feature request](https://github.com/inversify/InversifyJS/issues/new?assignees=rsaz&labels=needs+triage&projects=&template=feature_request.yml).

**Comment by Luluno01** [comment]
> Hey @Luluno01, I don't see how this is a bug. [Current docs](https://github.com/inversify/InversifyJS/blob/master/wiki/deactivation_handler.md) doesn't say the expected behavior is the one you suggests.
> 
> Having said that, we could talk about which behavior could make more sense. This issue is sort of confusing to me. In the first lines you state:
> 
> > The order of calling preDestroy hooks should be the reverse of the order of they are constructed. That is, we should destroy the leaves on the dependency tree first before we destroy their parents because leaves depend on parents. Destroying parents first breaks the dependencies of the leaves.
> 
> But later, in your example:
> 
> > `await container.unbindAllAsync()` should always call `preDestroy` hook on the instance of `ServiceB` first, wait until the hook resolves, then call `preDestroy` hook the instance of `ServiceA`.
> 
> In the context of a tree, I understand `ServiceB` is the parent node since `ServiceB` is allowed to have multiple children nodes. I understand parent nodes should be destroyed before children nodes, which makes total sense to me.
> 
> Did I understand you properly?
> 
> Either way, the current behavior does not enter in conflict with docs and therefore this is not an issue unless I'm missing something. I'll close the issue for now. If, after reading the docs you consider this is an issue, please mention me in this thread and I'll reopen it. Otherwise I would encourage you to open a [feature request](https://github.com/inversify/InversifyJS/issues/new?assignees=rsaz&labels=needs+triage&projects=&template=feature_request.yml).

TL; DR.: The definition of graph wasn't clear. You just confirmed this is a bug -- the actual behavior is `ServiceA` being destroyed first.

---

Thanks for your reply. I'm sorry I didn't clearly define what "dependency tree/graph" is. ~~By dependency tree/graph I mean a directed graph where an edge `parent => child` represents `child`'s runtime dependence on `parent`, not object reference graph.~~ I have updated the description with the well-accepted definition of "dependency graph".

~~Take my reproducer for example, by the definition above, the dependency graph comprises two nodes, `ServiceA` and `ServiceB` (here I use class names to represent their instances since we are talking about singletons), and one edge, `ServiceA => ServiceB`. This is interpreted as "`ServiceA` dominates `ServiceB`", or "`ServiceB` depends on `ServiceA`". The edge is created because `ServiceB` requires `ServiceA` to function (the `@inject(ServiceA)` part in `ServiceB`).~~

~~That said, if you are using an object reference graph instead of dependency graph~~, yes, `ServiceB` is the parent node. In such setup, your statement of "parent nodes should be destroyed before children nodes" is correct and that is also what I expect this library to do.

**Despite we all agree that in my reproducer, `ServiceB` should be destroyed first, it is not how it is currently implemented**. If you look at the comment right after `await container.unbindAllAsync()` in my reproducer, you will notice the actual behavior is the opposite of what we expect.

**Comment by Luluno01** [comment]
Here is the full output from `index.ts` for your reference:

> ServiceA constructor
> ServiceA init
> ServiceB constructor
> ServiceB init
> ServiceA destroy
> ServiceB destroy

**Comment by Luluno01** [comment]
Also, inter-dependent `preDestroy` hooks are actually called in parallel.

https://github.com/inversify/InversifyJS/blob/3761eb2182a11d2693692ebd344ba970d114a5a5/src/container/container.ts#L261

So the actual implementation is indeed destroying **all** singletons **in the same container** **at the same time** regardless of their dependency relationships.

**Comment by Luluno01** [comment]
Here is a piece of comment with regard to the behavior of the dependency injection feature of Spring for your reference:

https://github.com/spring-projects/spring-framework/blob/2e6c8daec639b8194bf191b9f67056cde2a18f48/spring-beans/src/main/java/org/springframework/beans/factory/support/DefaultSingletonBeanRegistry.java#L643C29-L644

**Comment by notaphplover** [comment]
Hey @Luluno01, as said, the only source of thruth for establishing what is the current behavior are inversify docs, which don't state the expected behavior is the one you describe.

As said, the behavior you propose might make sense, if you want the current behavior to be different than the actual one, please, open a feature request.

**Comment by Luluno01** [comment]
> Hey @Luluno01, as said, the only source of thruth for establishing what is the current behavior are inversify docs, which don't state the expected behavior is the one you describe.
> 
> As said, the behavior you propose might make sense, if you want the current behavior to be different than the actual one, please, open a feature request.

No problem. I have opened a feature request #1606. Somehow GitHub automatically assign that issue to you...
